Milliner/src/Service/MillinerService.php

```diff
@@ -20,230 +20,808 @@ class MillinerService implements MillinerServiceInterface
-                409
+                "Client error: `GET $fedora_url` resulted in a `$status $reason` response: " .
+                   $fedora_response->getBody(),
+                $status
             );
         }
 
-        $headers = [
-            'Authorization' => $token,
-            'Content-Type' => $mimetype,
-        ];
+        // Strip off the W/ prefix to make the ETag strong.
+        $etags = $fedora_response->getHeader("ETag");
+        $etag = ltrim(reset($etags), "W/");
+
+        // Get the modified date from the RDF.
+        $fedora_jsonld = json_decode(
+            $fedora_response->getBody(),
+            true
+        );
+
+        $fedora_modified = $this->getModifiedTimestamp(
+            $fedora_jsonld
+        );
+
+        // Get the jsonld from Drupal.
+        $headers = empty($token) ? [] : ['Authorization' => $token];
+        $drupal_response = $this->drupal->get(
+            $jsonld_url,
+            ['headers' => $headers]
+        );
+
+        $drupal_jsonld = json_decode(
+            $drupal_response->getBody(),
+            true
+        );
 
-        $fedora_response = $this->fedora->createResource(
-            '',
-            $drupal_binary,
+        // Mash it into the shape Fedora accepts.
+        $drupal_jsonld = $this->processJsonld(
+            $drupal_jsonld,
+            $jsonld_url,
+            $fedora_url
+        );
+
+        // Get the modified date from the RDF.
+        $drupal_modified = $this->getModifiedTimestamp(
+            $drupal_jsonld
+        );
+
+        // Abort with 412 if the Drupal RDF is stale.
+        if ($drupal_modified <= $fedora_modified) {
+            throw new \RuntimeException(
+                "Not updating $fedora_url because RDF at $jsonld_url is not newer",
+                412
+            );
+        }
+
+        // Conditional save it in Fedora.
+        $headers['Content-Type'] = 'application/ld+json';
+        $headers['Prefer'] = 'return=minimal; handling=lenient';
+        $headers['If-Match'] = $etag;
+        $response = $this->fedora->saveResource(
+            $fedora_url,
+            json_encode($drupal_jsonld),
             $headers
         );
 
-        $this->log->debug("Fedora POST Response: ", [
-            'body' => $fedora_response->getBody(),
-            'status' => $fedora_response->getStatusCode(),
-            'headers' => $fedora_response->getHeaders()
-        ]);
+        $status = $response->getStatusCode();
+        if (!in_array($status, [201, 204])) {
+            $reason = $response->getReasonPhrase();
+            throw new \RuntimeException(
+                "Client error: `PUT $fedora_url` resulted in a `$status $reason` response: " . $response->getBody(),
+                $status
+            );
+        }
+
+        // Map the URLS.
+        $this->gemini->saveUrls(
+            $uuid,
+            $jsonld_url,
+            $fedora_url,
+            $token
+        );
```

Comment:
I remember how/when I saved the urls was tricky because of all the different points of failure, but can't recall the _exact_ reason other than general paranoia.  Let me look back and see if I can figure out specifically why, because if not, we shouldn't.

Reply:
Yeah, I can't convince myself we need this.  Let's remove it and see what happens during testing.